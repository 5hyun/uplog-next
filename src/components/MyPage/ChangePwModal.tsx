import React, { useCallback, useEffect, useState } from 'react';
import {
  Flex,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay,
} from '@chakra-ui/react';
import { Link } from 'react-router-dom';
import { AiOutlineEye, AiOutlineEyeInvisible } from 'react-icons/ai';
import useInput from '../../hooks/useInput.ts';
import { useMessage } from '../../hooks/useMessage.ts';
import show = Mocha.reporters.Base.cursor.show;

interface Props {
  isOpen: boolean;
  onClose: () => void;
  isClickPwChange: boolean;
}
export default function ChangePwModal({ isOpen, onClose, isClickPwChange }: Props) {
  const [nowPw, onChangeNowPw, setNowPw] = useInput('');
  const [newPw, onChangeNewPw, setPassword] = useInput('');
  const [isCheckPw, setIsCheckPw] = useState(false);
  const [isPwVisible, setIsPwVisible] = useState(false);
  const { showMessage, contextHolder } = useMessage();

  const onClickChangePw = useCallback(() => {
    if (!nowPw || !newPw) {
      showMessage('warning', '비밀번호를 입력해주세요.');
      return;
    }

    onClose();
  }, []);
  const onClickPwVisible = useCallback(() => setIsPwVisible((prev) => !prev), []);

  const resetPw = useCallback(() => {
    setNowPw('');
    setPassword('');
    onClose();
  }, [onClose]);

  // 비밀번호 유효성 검사
  useEffect(() => {
    const regex = /^(?=.*[A-Za-z])(?=.*\d)(?=.*[!@#$%^&*])[A-Za-z\d!@#$%^&*]{8,15}$/;
    setIsCheckPw(regex.test(newPw));
  }, [newPw, isCheckPw]);

  return (
    <Modal isCentered onClose={resetPw} isOpen={isOpen}>
      <ModalOverlay />
      <ModalContent
        maxW="37.5rem"
        h={isClickPwChange ? '27rem' : '20rem'}
        shadow={'boxShadow-sign-up'}
        rounded={'none'}
        p={'1.2rem'}
        bg={'var(--white)'}
      >
        <ModalHeader
          fontSize={'1.8rem'}
          fontWeight={700}
          bg={'var(--white)'}
          color={'var(--black)'}
        >
          {isClickPwChange ? '비밀번호 변경' : '계정 삭제'}
        </ModalHeader>
        <ModalCloseButton
          fontSize={'1rem'}
          color={'var(--gray-light)'}
          mt={'0.6rem'}
          mr={'0.8rem'}
        />
        <ModalBody>
          <Flex justifyContent={'center'} h={'100%'}>
            <section className={'flex-col-center justify-evenly w-[22.5rem] h-full'}>
              {contextHolder}

              {/* 현재 비밀번호 */}
              <div className={'w-full text-[0.93rem] mb-4'}>
                <span className={'text-gray-dark font-bold mb-[0.93rem]'}>현재 비밀번호</span>
                <input
                  type="password"
                  value={nowPw}
                  onChange={onChangeNowPw}
                  placeholder={'현재 비밀번호를 입력하세요.'}
                  maxLength={15}
                  className={
                    'tracking-[-0.1rem] w-full h-10 border-base border-gray-border rounded-xl mb-2 p-4 text-black'
                  }
                />
                {isClickPwChange && (
                  <div className={'flex-row-center justify-start pl-4 text-sm text-gray-light'}>
                    <span>비밀번호를 잊으셨나요?</span>
                    <Link to={'/pwinquiry'} className={'ml-2 underline'}>
                      비밀번호 찾기
                    </Link>
                  </div>
                )}
              </div>
              {/* 새로운 비밀번호 */}
              {isClickPwChange && (
                <div className={'w-full text-[0.93rem]'}>
                  <span className={'text-gray-dark font-bold mb-[0.93rem]'}>새로운 비밀번호</span>
                  <span
                    className={
                      'flex-row-center border-base border-gray-border rounded-xl h-10 pl-4 mb-2 bg-border'
                    }
                  >
                    <input
                      type={`${isPwVisible ? 'text' : 'password'}`}
                      value={newPw}
                      onChange={onChangeNewPw}
                      placeholder={'새로운 비밀번호를 입력하세요.'}
                      maxLength={15}
                      className={`w-[90%] h-full text-black
                    ${newPw && !isPwVisible && 'tracking-[-0.1rem]'}
                    `}
                    />
                    <span className={'w-[10%] text-2xl cursor-pointer'} onClick={onClickPwVisible}>
                      {isPwVisible ? (
                        <AiOutlineEyeInvisible className={'fill-gray-light'} />
                      ) : (
                        <AiOutlineEye className={'fill-gray-light'} />
                      )}
                    </span>
                  </span>
                  {!isCheckPw && newPw && (
                    <span className={'flex-row-center justify-start text-sm text-[#E06469]'}>
                      영문/숫자/특수문자 포함, 8~15자로 입력해주세요.
                    </span>
                  )}
                </div>
              )}
            </section>
          </Flex>
        </ModalBody>

        <ModalFooter>
          <button
            className={`bg-orange rounded font-bold text-xs text-white
            ${isClickPwChange ? 'w-[4.5rem] h-9' : 'w-[6rem] h-9'}`}
            onClick={onClickChangePw}
          >
            {isClickPwChange ? '확인' : '계정 삭제'}
          </button>
        </ModalFooter>
      </ModalContent>
    </Modal>
  );
}
