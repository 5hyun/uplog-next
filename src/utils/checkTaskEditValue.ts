import { TaskData } from '@/typings/task.ts';

export function checkTaskEditValue(editTask: TaskData): boolean {
  if (editTask.taskName === '') {
    return false;
  }
  if (editTask.targetMemberInfoDTO && editTask.targetMemberInfoDTO.id === 0) {
    return false;
  }
  if (editTask.menuId === 0) {
    return false;
  }
  if (editTask.teamId === 0) {
    return false;
  }
  if (editTask.startTime === '' || editTask.endTime === '') {
    return false;
  }

  return true;
}
