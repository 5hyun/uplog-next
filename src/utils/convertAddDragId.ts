import { StatusTaskData, Tasks } from '@/typings/task.ts';
import { useState } from 'react';

export function convertAddDragId(tasks: StatusTaskData): Tasks {
  const [convertedTasks, setConvertedTasks] = useState<Tasks>({
    PROGRESS_BEFORE: [],
    PROGRESS_IN: [],
    PROGRESS_COMPLETE: [],
  });

  const beforeStatusTasks = tasks.PROGRESS_BEFORE.map((task) => ({
    ...task,
    dragId: task.id.toString(), // dragId 값을 문자열로 변환하여 할당
  }));

  setConvertedTasks({
    ...convertedTasks,
    PROGRESS_BEFORE: beforeStatusTasks,
  });

  const inStatusTasks = tasks.PROGRESS_IN.map((task) => ({
    ...task,
    dragId: task.id.toString(), // dragId 값을 문자열로 변환하여 할당
  }));

  setConvertedTasks({
    ...convertedTasks,
    PROGRESS_IN: inStatusTasks,
  });

  const completeStatusTasks = tasks.PROGRESS_COMPLETE.map((task) => ({
    ...task,
    dragId: task.id.toString(), // dragId 값을 문자열로 변환하여 할당
  }));

  setConvertedTasks({
    ...convertedTasks,
    PROGRESS_COMPLETE: completeStatusTasks,
  });

  return convertedTasks;
}
