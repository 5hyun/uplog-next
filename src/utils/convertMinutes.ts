// ** 타이머 분:초 구하는 함수 */
export function convertMinutes(seconds: number): string {
  const minutes = Math.floor(seconds / 60);
  const remainingSeconds = seconds % 60;
  const formattedSeconds = remainingSeconds < 10 ? `0${remainingSeconds}` : `${remainingSeconds}`;

  if (minutes === 0) {
    return `00:${formattedSeconds}`;
  }

  return `${minutes}:${formattedSeconds}`;
}
