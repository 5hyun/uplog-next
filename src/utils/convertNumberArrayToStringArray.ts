export function convertNumberArrayToStringArray(numbers: number[]): string[] {
  // 배열의 각 요소를 문자열로 변환하여 새로운 배열을 생성합니다.
  const strings: string[] = numbers.map((number) => number.toString());
  return strings;
}
