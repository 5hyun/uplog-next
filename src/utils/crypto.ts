import CryptoJS from 'crypto-js';

/* Salt 비밀 키 */
// const salt = import.meta.env.VITE_SALT;
const salt = '1234';

/* 암호화 */
export const encrypt = (text: string) => {
  // 값이 없을 경우 빈 문자열 반환
  if (!text) return '';
  return CryptoJS.AES.encrypt(text, salt).toString();
};

/* 복호화 */
export const decrypt = (text: string) => {
  // 값이 없을 경우 빈 문자열 반환
  if (!text) return '';

  try {
    const bytes = CryptoJS.AES.decrypt(text, salt); // 복호화 시도
    return bytes.toString(CryptoJS.enc.Utf8);
  } catch (error) {
    return ''; // 에러 발생 시 빈 문자열 반환
  }
};
