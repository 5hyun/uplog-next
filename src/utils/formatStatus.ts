// ** 진행 전,중,후 상태로 변환하기 위한 함수 */

export function formatStatus(status: string): string {
  if (status === 'PROGRESS_BEFORE') {
    return '진행 전';
  } else if (status === 'PROGRESS_IN') {
    return '진행 중';
  }

  return '진행 후';
}
