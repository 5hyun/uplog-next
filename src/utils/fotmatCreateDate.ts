export function formatCreateDate(date: string) {
  const options: Intl.DateTimeFormatOptions = {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
    hour: 'numeric',
    minute: 'numeric',
    hour12: true,
    timeZone: 'Asia/Seoul',
  };

  const formattedDate = new Date(date).toLocaleString('ko-KR', options);

  return formattedDate;
}
