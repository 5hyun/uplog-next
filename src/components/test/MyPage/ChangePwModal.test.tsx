import { describe, expect, test } from '@jest/globals';
import { fireEvent, getByText, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { withRouter } from '../../../tests/utils';
import { Route } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import ChangePwModal from '../../MyPage/ChangePwModal';
import userEvent from '@testing-library/user-event';

describe('ChangePwModal', () => {
  const misPw = 'qwer234';
  const goodPw = 'qwer234!';

  const onClose = jest.fn();

  const props_pwChange = {
    isOpen: true,
    onClose: onClose,
    isClickPwChange: true,
  };
  const props_memberWithdrawal = {
    isOpen: true,
    onClose: onClose,
    isClickPwChange: false,
  };

  // 테스트가 끝날때 마다 Mock 함수를 깨끗하게 정리
  afterEach(() => onClose.mockReset());

  // 비밀번호 변경 모달
  test('password change modal correctly', () => {
    const component = renderer.create(
      withRouter(<Route path={'/'} element={<ChangePwModal {...props_pwChange} />} />, '/mypage')
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('password change modal, test password', () => {
    render(
      withRouter(
        <Route path={'/mypage'} element={<ChangePwModal {...props_pwChange} />} />,
        '/mypage'
      )
    );

    const passwordInput = screen.getByPlaceholderText('새로운 비밀번호를 입력하세요.');

    fireEvent.change(passwordInput, { target: { value: misPw } });
    expect(screen.getByText('영문/숫자/특수문자 포함, 8~15자로 입력해주세요.')).toBeInTheDocument();

    fireEvent.change(passwordInput, { target: { value: goodPw } });
    expect(
      screen.queryByText('영문/숫자/특수문자 포함, 8~15자로 입력해주세요.')
    ).not.toBeInTheDocument();
  });

  test('move page, find password button click to /pwinquiry', () => {
    render(
      withRouter(
        <>
          <Route path={'/mypage'} element={<ChangePwModal {...props_pwChange} />} />
          <Route path={'/pwinquiry'} element={<p>비밀번호 찾기 페이지</p>} />
        </>,
        '/mypage'
      )
    );

    const findPwLink = screen.getByRole('link', { name: '비밀번호 찾기' });
    fireEvent.click(findPwLink);
    expect(screen.getByText('비밀번호 찾기 페이지')).toBeInTheDocument();
  });

  // 회원 탈퇴 모달
  test('Member Withdrawal modal correctly', () => {
    const component = renderer.create(
      withRouter(
        <Route path={'/mypage'} element={<ChangePwModal {...props_memberWithdrawal} />} />,
        'mypage'
      )
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('check Withdrawal modal screen', () => {
    render(
      withRouter(
        <Route path={'/mypage'} element={<ChangePwModal {...props_memberWithdrawal} />} />,
        '/mypage'
      )
    );

    expect(screen.getByText('현재 비밀번호')).toBeInTheDocument();
    // getByText는 존재하지 않으명 에러가 나는데, queryByText는 존재하지 않으면 애러가 안나서 not을 하고 싶으면 queryByT
    expect(screen.queryByText('새로운 비밀번호')).not.toBeInTheDocument();
  });
});
