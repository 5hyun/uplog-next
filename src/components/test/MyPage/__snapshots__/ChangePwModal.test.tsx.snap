// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePwModal Member Withdrawal modal correctly 1`] = `null`;

exports[`ChangePwModal password change modal correctly 1`] = `null`;
